drivers/timer/cortex_m_systick.c

```diff
@@ -201,8 +201,8 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
 		delay =
 		 ((delay + CYC_PER_TICK - 1) / CYC_PER_TICK) * CYC_PER_TICK;
 		delay -= unannounced;
-		delay = MAX(delay, MIN_DELAY);
-		if (delay > MAX_CYCLES) {
+		delay = MAX(delay, (uint32_t)MIN_DELAY);
+		if (delay > (uint32_t)MAX_CYCLES) {
```

Comment:
same here. the definition itself ought to be unsigned.

Reply:
There are cascading effects of changing this MACRO to be unsigned.

```
/home/blues/zephyrproject/zephyr/drivers/timer/cortex_m_systick.c:173:58: warning: operand of '?:' changes signedness from 'int32_t' {aka 'int'} to 'unsigned int' due to unsignedness of other operand [-Wsign-compare]
  173 |         ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
      |                                                          ^~~~~
```

Where `ticks` is a parameter in the current function:
```
void sys_clock_set_timeout(int32_t ticks, bool idle)
{
```

Then we would need to change the function definition to become `uint32_t`.

However, this seems to be okay, as the function does contain a `CLAMP` macro that appears to force the value to be greater than or equal to zero. I'll go ahead and update the full scope of the cascading effects, and you will be able to see if you like it in the next commit.

Reply:
Upon further investigation, the documentation appears to be wrong (or misleading).

Specifically...

> Similarly a ticks value of zero (or even negative) is legal and
> treated identically: it simply indicates the kernel would like the
> next tick announcement as soon as possible.

**`drivers/timer/cortex_m_systick.c:162`**
```
if (IS_ENABLED(CONFIG_TICKLESS_KERNEL) && idle && ticks == K_TICKS_FOREVER) {
```

It seems like the comparison should be whether or not `ticks <= 0`, because `K_TICKS_FOREVER` is defined as `-1`.

**`include/zephyr/sys_clock.h:45-51`**
```
#ifdef CONFIG_TIMEOUT_64BIT
typedef int64_t k_ticks_t;
#else
typedef uint32_t k_ticks_t;
#endif

#define K_TICKS_FOREVER ((k_ticks_t) -1)
```

Later in the code, they have attempted to compensate for any negative number by performing the following transformation.

**`drivers/timer/cortex_m_systick.c:173-174`**
```
	ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
	ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
```

However, this results in `-1` (`K_TICKS_FOREVER`) being treated differently than -2, which will become zero.

This definitely appears to be a bug, but I don't know how it should be handled. :shrug:

Reply:
I will need input from the team before I know how it should be handled. :shrug:

Reply:
After stepping away for a bit, I realized I can move forward and leave the exact same problematic logic in place.

I just need to change the result of the CLAMP macro to an unsigned variable, and propagate the variable internally throughout the function.

Reply:
It's not as easy to bypass as I'd hoped.

When the value of `MAX_TICKS` is sufficiently large, then the outcome of the logic in the following two lines...

```cpp
	ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
	ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
```

... is heavily dependent on the "signedness" of the integer. Which is why I expect the bounding logic looks so bizarre.

In plain English, I would suspect the intent of the function is to preserve the special meaning of `K_TICKS_FOREVER`, otherwise bound `ticks` between `0` and `MAX_TICKS`.

Let's look at following scenarios, remembering `ticks` is a signed value, `K_TICKS_FOREVER` is -1, and pretending that `k_ticks_t` is a `uint8_t` for ease of reading...

1. **ticks == 75 and MAX_TICKS == 100**
   1. `ticks` remains 75  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes 74  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
2. **ticks == 127 and MAX_TICKS == 100**
   1. `ticks` remains 127  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes 100  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
3. **ticks == -1 and MAX_TICKS == 100**
   1. `ticks` becomes 100  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes 99  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
4. **ticks == -1 and MAX_TICKS == 200**
   1. `ticks` becomes -56 (i.e. 200)  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes 0  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
5. **ticks == -2 and MAX_TICKS == 200**
   1. `ticks` remains -2  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes 0  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
6. **ticks == 100 and MAX_TICKS == 200**
   1. `ticks` remains 100  // ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
   2. `ticks` becomes -56  // ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);

As you can see, this logic leads to some pretty wild and unexpected outcomes.

Reply:
@stephanosio, I could really use your guidance here.